llama-server: add router multi-model tests (#17704)

ServeurpersoCom opened the pull request:
Add 4 test cases for model router:
- test_router_unload_model: explicit model unloading
- test_router_models_max_evicts_lru: LRU eviction with --models-max
- test_router_no_models_autoload: --no-models-autoload flag behavior
- test_router_api_key_required: API key authentication

Tests use async model loading with polling and graceful skip when insufficient models available for eviction testing.

utils.py changes:
- Add models_max, models_dir, no_models_autoload attributes to ServerProcess
- Handle JSONDecodeError for non-JSON error responses (fallback to text)

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

Close #17704

ServeurpersoCom committed:
llama-server: add router multi-model tests (#17704)

Add 4 test cases for model router:
- test_router_unload_model: explicit model unloading
- test_router_models_max_evicts_lru: LRU eviction with --models-max
- test_router_no_models_autoload: --no-models-autoload flag behavior
- test_router_api_key_required: API key authentication

Tests use async model loading with polling and graceful skip when
insufficient models available for eviction testing.

utils.py changes:
- Add models_max, models_dir, no_models_autoload attributes to ServerProcess
- Handle JSONDecodeError for non-JSON error responses (fallback to text)

ServeurpersoCom committed:
llama-server: update test models to new HF repos

ngxson commented:
I mirrored the PR here for faster CI: https://github.com/ngxson/llama.cpp/pull/48

ngxson commented on tools/server/tests/unit/test_router.py:
```diff
@@ -48,3 +48,152 @@ def test_router_chat_completion_stream(model: str, success: bool):
+        "ggml-org/tinygemma3-GGUF:Q8_0",
+        "ggml-org/models/tinyllamas/stories260K.gguf",
+        "ggml-org/models/bert-bge-small/ggml-model-f16.gguf",
+    ]
```
I think this need to be updated too

ngxson reviewed (approved):
Great jobs, thanks! Merging once CI passes

ServeurpersoCom commented:
On empty tmp/* I check this after eating :

FAILED unit/test_router.py::test_router_models_max_evicts_lru - AssertionError: Timed out waiting for ggml-org/tinygemma3-GGUF:Q8_0 to reach {'loaded'}, last status: unloaded

ngxson committed:
add offline

ServeurpersoCom committed:
llama-server: fix router LRU eviction test and add preloading

Fix eviction test: load 2 models first, verify state, then load
3rd to trigger eviction. Previous logic loaded all 3 at once,
causing first model to be evicted before verification could occur.

Add module fixture to preload models via ServerPreset.load_all()
and mark test presets as offline to use cached models

ngxson commented:
Can be merged once the CI https://github.com/ngxson/llama.cpp/pull/48 passes

ServeurpersoCom committed:
llama-server: fix split model download on Windows
